cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/common/AbstractStageContext.java

```diff
@@ -192,51 +192,99 @@ public URL call() {
 
   @Override
   public Map<MetadataScope, Metadata> getMetadata(MetadataEntity metadataEntity) {
-    return null;
+    return CALLER.callUnchecked(() -> pipelineRuntime.getMetadataReader()
+      .orElseThrow(this::createMetadataUnsupportedException)
+      .getMetadata(metadataEntity));
   }
 
   @Override
   public Metadata getMetadata(MetadataScope scope, MetadataEntity metadataEntity) {
-    return null;
+    return CALLER.callUnchecked(() -> pipelineRuntime.getMetadataReader()
+      .orElseThrow(this::createMetadataUnsupportedException)
+      .getMetadata(scope, metadataEntity));
   }
 
   @Override
   public void addProperties(MetadataEntity metadataEntity, Map<String, String> properties) {
-
+    CALLER.callUnchecked((Callable<Void>) () -> {
+      pipelineRuntime.getMetadataWriter()
+        .orElseThrow(this::createMetadataUnsupportedException)
+        .addProperties(metadataEntity, properties);
+      return null;
+    });
   }
 
   @Override
   public void addTags(MetadataEntity metadataEntity, String... tags) {
-
+    CALLER.callUnchecked((Callable<Void>) () -> {
+      pipelineRuntime.getMetadataWriter()
+        .orElseThrow(this::createMetadataUnsupportedException)
+        .addTags(metadataEntity, tags);
+      return null;
+    });
   }
 
   @Override
   public void addTags(MetadataEntity metadataEntity, Iterable<String> tags) {
-
+    CALLER.callUnchecked((Callable<Void>) () -> {
+      pipelineRuntime.getMetadataWriter()
```

Comment:
Just noticed how repetitive it is. Maybe better to have private methods

`private MetadataReader getMetadataReader()` and `private MetadataWriter getMetadataWriter()`